File: src/translate_video/core/schemas.py
"""Сериализуемые схемы проекта, используемые во всем движке."""

from __future__ import annotations

from dataclasses import asdict, dataclass, field
from datetime import UTC, datetime
from enum import StrEnum
from pathlib import Path
from typing import Any
from uuid import uuid4

from translate_video.core.config import PipelineConfig


class SegmentStatus(StrEnum):
    """Состояния жизненного цикла одного текстового/аудио сегмента."""

    DRAFT = "draft"
    TRANSCRIBED = "transcribed"
    TRANSLATED = "translated"
    TTS_READY = "tts_ready"
    FAILED = "failed"


class ProjectStatus(StrEnum):
    """Состояния жизненного цикла одного проекта перевода."""

    CREATED = "created"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class Stage(StrEnum):
    """Этапы пайплайна для задач, артефактов и webhook-событий."""

    INIT = "init"
    PROBE = "probe"
    EXTRACT_AUDIO = "extract_audio"
    TRANSCRIBE = "transcribe"
    REGROUP = "regroup"     # перегруппировка по предложениям (TVIDEO-039)
    TRANSLATE = "translate"
    VOICE_CAST = "voice_cast"
    TTS = "tts"
    TIMING_FIT = "timing_fit"
    MIX = "mix"
    RENDER = "render"
    QA = "qa"
    EXPORT = "export"


class ArtifactKind(StrEnum):
    """Стабильные категории артефактов внутри папки проекта."""

    SETTINGS = "settings"
    SOURCE_AUDIO = "source_audio"
    SOURCE_TRANSCRIPT = "source_transcript"
    TRANSLATED_TRANSCRIPT = "translated_transcript"
    SPEAKERS = "speakers"
    SUBTITLES = "subtitles"
    TTS_AUDIO = "tts_audio"
    FINAL_AUDIO = "final_audio"
    OUTPUT_VIDEO = "output_video"
    QA_REPORT = "qa_report"


class JobStatus(StrEnum):
    """Статус одного перезапускаемого выполнения этапа."""

    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    SKIPPED = "skipped"


@dataclass(slots=True)
class ArtifactRecord:
    """Типизированные метаданные артефакта с путем относительно проекта."""

    kind: ArtifactKind
    path: str
    stage: Stage
    content_type: str = "application/json"
    created_at: str = field(default_factory=lambda: datetime.now(UTC).isoformat())
    metadata: dict[str, Any] = field(default_factory=dict)
    checksum: str | None = None

    def to_dict(self) -> dict[str, Any]:
        """Вернуть JSON-совместимую запись артефакта."""

        return asdict(self)

    @classmethod
    def from_dict(cls, payload: dict[str, Any]) -> "ArtifactRecord":
        """Создать запись артефакта и восстановить enum-значения."""

        return cls(
            kind=ArtifactKind(payload["kind"]),
            path=payload["path"],
            stage=Stage(payload["stage"]),
            content_type=payload.get("content_type", "application/json"),
            created_at=payload.get("created_at", datetime.now(UTC).isoformat()),
            metadata=dict(payload.get("metadata", {})),
            checksum=payload.get("checksum"),
        )


@dataclass(slots=True)
class StageRun:
    """Одна попытка выполнения этапа пайплайна."""

    stage: Stage
    status: JobStatus = JobStatus.PENDING
    id: str = field(default_factory=lambda: f"stage_{uuid4().hex[:12]}")
    started_at: str | None = None
    finished_at: str | None = None
    inputs: list[str] = field(default_factory=list)
    outputs: list[str] = field(default_factory=list)
    error: str | None = None
    attempt: int = 1
    progress_current: int | None = None
    progress_total: int | None = None
    progress_message: str | None = None

    def to_dict(self) -> dict[str, Any]:
        """Вернуть JSON-совместимую запись запуска этапа."""

        return asdict(self)

    @classmethod
    def from_dict(cls, payload: dict[str, Any]) -> "StageRun":
        """Создать запись запуска этапа и восстановить enum-значения."""

        return cls(
            id=payload.get("id", f"stage_{uuid4().hex[:12]}"),
            stage=Stage(payload["stage"]),
            status=JobStatus(payload.get("status", "pending")),
            started_at=payload.get("started_at"),
            finished_at=payload.get("finished_at"),
            inputs=list(payload.get("inputs", [])),
            outputs=list(payload.get("outputs", [])),
            error=payload.get("error"),
            attempt=payload.get("attempt", 1),
            progress_current=payload.get("progress_current"),
            progress_total=payload.get("progress_total"),
            progress_message=payload.get("progress_message"),
        )


@dataclass(slots=True)
class Segment:
    """Один речевой сегмент и состояние его перевода."""

    start: float
    end: float
    source_text: str
    id: str = field(default_factory=lambda: f"seg_{uuid4().hex[:12]}")
    translated_text: str = ""
    speaker_id: str | None = None
    voice: str | None = None
    confidence: float | None = None
    status: SegmentStatus = SegmentStatus.DRAFT
    tts_path: str | None = None
    tts_text: str = ""  # текст реально отправленный в TTS (пусто = translated_text)
    qa_flags: list[str] = field(default_factory=list)

    def __post_init__(self) -> None:
        if self.end < self.start:
            raise ValueError("конец сегмента должен быть больше или равен началу")
        self.status = SegmentStatus(self.status)

    @property
    def duration(self) -> float:
        """Длительность сегмента в секундах."""

        return self.end - self.start

    def to_dict(self) -> dict[str, Any]:
        """Вернуть JSON-совместимое представление сегмента."""

        return asdict(self)

    @classmethod
    def from_dict(cls, payload: dict[str, Any]) -> "Segment":
        """Создать сегмент из JSON-данных."""

        data = dict(payload)
        data["status"] = SegmentStatus(data.get("status", "draft"))
        return cls(**data)


@dataclass(slots=True)
class VideoProject:
    """Сохраняемые метаданные одного запуска перевода."""

    input_video: Path
    work_dir: Path
    config: PipelineConfig
    id: str = field(default_factory=lambda: f"project_{uuid4().hex[:12]}")
    segments: list[Segment] = field(default_factory=list)
    artifacts: dict[str, str] = field(default_factory=dict)
    artifact_records: list[ArtifactRecord] = field(default_factory=list)
    stage_runs: list[StageRun] = field(default_factory=list)
    status: ProjectStatus = ProjectStatus.CREATED
    # Снапшоты баланса до/после пайплайна — для вычисления реального расхода.
    # Ключи: "polza_before", "polza_after", "neuroapi_before", "neuroapi_after"
    billing_snapshots: dict[str, float] = field(default_factory=dict)

    def __post_init__(self) -> None:
        self.status = ProjectStatus(self.status)

    def to_dict(self) -> dict[str, Any]:
        """Вернуть JSON-совместимое представление проекта."""

        return {
            "id": self.id,
            "input_video": str(self.input_video),
            "work_dir": str(self.work_dir),
            "config": self.config.to_dict(),
            "segments": [segment.to_dict() for segment in self.segments],
            "artifacts": self.artifacts,
            "artifact_records": [record.to_dict() for record in self.artifact_records],
            "stage_runs": [run.to_dict() for run in self.stage_runs],
            "status": self.status,
            "billing_snapshots": self.billing_snapshots,
        }

    @classmethod
    def from_dict(cls, payload: dict[str, Any]) -> "VideoProject":
        """Создать проект из JSON-данных."""

        return cls(
            id=payload["id"],
            input_video=Path(payload["input_video"]),
            work_dir=Path(payload["work_dir"]),
            config=PipelineConfig.from_dict(payload["config"]),
            segments=[Segment.from_dict(item) for item in payload.get("segments", [])],
            artifacts=dict(payload.get("artifacts", {})),
            artifact_records=[
                ArtifactRecord.from_dict(item) for item in payload.get("artifact_records", [])
            ],
            stage_runs=[StageRun.from_dict(item) for item in payload.get("stage_runs", [])],
            status=ProjectStatus(payload.get("status", "created")),
            billing_snapshots=dict(payload.get("billing_snapshots", {})),
        )

File: ui/src/types/schemas.ts
export type SegmentStatus = "draft" | "transcribed" | "translated" | "tts_ready" | "failed";
export type ProjectStatus = "created" | "running" | "completed" | "failed";
export type JobStatus = "pending" | "running" | "completed" | "failed" | "skipped";
export type Stage = "init" | "probe" | "extract_audio" | "transcribe" | "regroup" | "translate" | "voice_cast" | "tts" | "timing_fit" | "mix" | "render" | "qa" | "export";

export interface PipelineConfig {
    source_language: string;
    target_language: string;
    translation_mode: string;
    translation_quality: string;
    translation_style: string;
    adaptation_level: string;
    voice_strategy: string;
    quality_gate: string;
    terminology_domain: string;
    target_audience: string;
    profanity_policy: string;
    units_policy: string;
    preserve_names: boolean;
    preserve_brand_names: boolean;
    original_audio_volume: number;
    background_ducking: boolean;
    subtitle_formats: string[];
    subtitle_embed_mode?: string;  // 'none' | 'soft' | 'burn'
    glossary_path?: string;
    timing_policy: string;
    target_chars_per_second: number;
    timing_fit_max_rewrites: number;
    use_cloud_timing_rewriter: boolean;
    rewrite_provider_order: string[];
    use_cloud_translation: boolean;
    translation_provider_order: string[];
    translation_provider_timeout: number;
    translation_allow_paid_fallback: boolean;
    professional_translation_provider: string;
    professional_translation_model: string;
    professional_rewrite_provider: string;
    professional_rewrite_model: string;
    professional_tts_provider: string;
    professional_tts_model: string;
    professional_tts_voice: string;
    professional_tts_voice_2: string;
    professional_tts_role: string;
    professional_tts_role_2: string;
    professional_tts_speed: number;
    professional_tts_speed_2: number;
    professional_tts_pitch: number;
    professional_tts_pitch_2: number;
    professional_tts_emotion: number;
    el_stability: number;
    el_similarity_boost: number;
    el_style: number;
    el_speed: number;
    allow_tts_rate_adaptation: boolean;
    allow_render_audio_speedup: boolean;
    allow_timeline_shift: boolean;
    max_timeline_shift: number;
    tts_base_rate: number;
    tts_max_rate: number;
    tts_rate_slack: number;
    render_max_speed: number;
    render_gap: number;
    allow_render_audio_trim: boolean;
    regroup_max_slot: number;
    do_not_translate: string[];
    dev_mode: boolean;
}

export type PipelineConfigDraft = Partial<PipelineConfig>;

export interface Segment {
    id: string;
    start: number;
    end: number;
    source_text: string;
    translated_text: string;
    speaker_id?: string;
    voice?: string;
    confidence?: number;
    status: SegmentStatus;
    tts_path?: string;
    tts_text?: string;
    qa_flags?: string[];
}

export interface StageRun {
    id: string;
    stage: Stage;
    status: JobStatus;
    started_at?: string;
    finished_at?: string;
    inputs: string[];
    outputs: string[];
    error?: string;
    attempt: number;
    progress_current?: number | null;
    progress_total?: number | null;
    progress_message?: string | null;
    metadata?: Record<string, unknown>;
}

export interface ArtifactRecord {
    kind: string;
    path: string;
    stage: Stage;
    content_type: string;
    created_at: string;
    metadata: Record<string, unknown>;
    checksum?: string;
}

export interface VideoProject {
    project_id: string;
    status: ProjectStatus;
    input_video: string;
    work_dir: string;
    segments: number | Segment[];
    artifacts: Record<string, string>;
    artifact_records?: ArtifactRecord[];
    stage_runs?: StageRun[];
    config?: PipelineConfig;
    /** Прогресс выполнения 0–100 */
    progress_percent?: number | null;
    /** ETA до завершения в секундах */
    eta_seconds?: number | null;
    /** ISO-дата начала пайплайна */
    started_at?: string | null;
    /** Сообщение об ошибке (для failed статуса) */
    error?: string | null;
}

export interface ProjectListResponse {
    projects: VideoProject[];
}

export interface ArtifactsResponse {
    project_id: string;
    work_dir: string;
    artifacts: ArtifactRecord[];
}

export interface PreflightCheck {
    name: string;
    ok: boolean;
    message: string;
    details: Record<string, string>;
}

export interface CostEstimate {
    translation_usd: number;
    tts_usd: number;
    total_usd: number;
    currency: string;
    note: string;
}

export interface PreflightReport {
    input_video: string;
    provider: string;
    ok: boolean;
    duration_seconds?: number | null;
    /** Оценка стоимости обработки */
    cost_estimate?: CostEstimate | null;
    /** ETA всего пайплайна в секундах */
    duration_estimate_seconds?: number | null;
    checks: PreflightCheck[];
}

export interface ProviderModel {
    id: string;
    name: string;
}

export interface ProviderBalance {
    provider: string;
    configured: boolean;
    balance?: number | null;
    currency?: string | null;
    used?: number | null;
    source?: string | null;
    message?: string | null;
}
